SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/OwnershipLockingPolicy.cpp

```diff
@@ -0,0 +1,296 @@
+	bool ReleaseSucceeded = ReleaseLock(LockToken);
+	return ReleaseSucceeded;
+}
+
+void UOwnershipLockingPolicy::OnOwnerUpdated(const AActor* Actor, const AActor* OldOwner)
+{
+	check(Actor != nullptr);
+
+	// If an explicitly locked Actor is changing owner.
+	if (IsExplicitlyLocked(Actor))
+	{
+		RecalculateLockedActorOwnershipHierarchyInformation(Actor);
+	}
+
+	// If a locked hierarchy root is changing owner.
+	if (IsLockedHierarchyRoot(Actor))
+	{
+		RecalculateAllExplicitlyLockedActorsInThisHierarchy(Actor);
+	}
+	// If an Actor in a locked hierarchy is changing owner (i.e. either the old owner or
+	// the root hierarchy of the old owner)is the root of a locked hierarchy), we need to
+	// recalculate ownership hierarchies of all explicitly locked Actors in that hierarchy.
+	else if (OldOwner != nullptr)
+	{
+		if (const AActor* OldHierarchyRoot = IsLockedHierarchyRoot(OldOwner) ? OldOwner : SpatialGDK::GetHierarchyRoot(OldOwner))
+		{
+			RecalculateAllExplicitlyLockedActorsInThisHierarchy(OldHierarchyRoot);
+		}
+	}
+}
+
+void UOwnershipLockingPolicy::OnExplicitlyLockedActorDeleted(AActor* DestroyedActor)
+{
+	// Find all tokens for this Actor and unlock.
+	for (auto TokenNameActorIterator = TokenToNameAndActor.CreateIterator(); TokenNameActorIterator; ++TokenNameActorIterator)
+	{
+		if (TokenNameActorIterator->Value.Actor == DestroyedActor)
+		{
+			TokenNameActorIterator.RemoveCurrent();
+		}
+	}
+
+	// Delete Actor from local mapping.
+	MigrationLockElement ActorLockingState = ActorToLockingState.FindAndRemoveChecked(DestroyedActor);
+
+	// Update ownership path Actor mapping to remove this Actor.
+	RemoveOwnershipHierarchyRootInformation(ActorLockingState.HierarchyRoot, DestroyedActor);
+}
+
+void UOwnershipLockingPolicy::OnHierarchyRootActorDeleted(AActor* DestroyedOwnershipPathActor)
+{
+	check(LockedOwnershipRootActorToExplicitlyLockedActors.Contains(DestroyedOwnershipPathActor));
+
+	// For all explicitly locked Actors where this Actor is on the ownership path, recalculate the
+	// ownership path information to account for this Actor's deletion.
+	RecalculateAllExplicitlyLockedActorsInThisHierarchy(DestroyedOwnershipPathActor);
+	LockedOwnershipRootActorToExplicitlyLockedActors.Remove(DestroyedOwnershipPathActor);
+}
+
+void UOwnershipLockingPolicy::RecalculateAllExplicitlyLockedActorsInThisHierarchy(const AActor* HierarchyRoot)
+{
+	TArray<const AActor*> ExplicitlyLockedActorsWithThisActorInOwnershipPath = LockedOwnershipRootActorToExplicitlyLockedActors.FindChecked(HierarchyRoot);
+	for (const AActor* ExplicitlyLockedActor : ExplicitlyLockedActorsWithThisActorInOwnershipPath)
+	{
+		RecalculateLockedActorOwnershipHierarchyInformation(ExplicitlyLockedActor);
+	}
+}
+
+void UOwnershipLockingPolicy::RecalculateLockedActorOwnershipHierarchyInformation(const AActor* ExplicitlyLockedActor, const AActor* DeletedHierarchyActor)
```

Comment:
This is never called with a valid DeletedHierarchyActor anymore, I think all this functionality can be greatly simplified.